lib/plugins/package/lib/packageService.js

```diff
@@ -59,44 +61,115 @@ module.exports = {
+      }
+
+      // Enqueue the next item in the array
+      return concurrencyPromise.then(() => enqueue());
+    };
+
+    return enqueue().then(() => Promise.all(returnValues));
+  },
+
   packageService() {
     this.serverless.cli.log('Packaging service...');
     let shouldPackageService = false;
     const allFunctions = this.serverless.service.getAllFunctions();
-    let packagePromises = allFunctions.map(functionName => {
+    const allLayers = this.serverless.service.getAllLayers();
+
+    const packageFunction = functionName => {
       const functionObject = this.serverless.service.getFunction(functionName);
       functionObject.package = functionObject.package || {};
       if (functionObject.package.disable) {
         this.serverless.cli.log(`Packaging disabled for function: "${functionName}"`);
-        return BbPromise.resolve();
+        return Promise.resolve();
       }
       if (functionObject.package.artifact) {
-        return BbPromise.resolve();
+        return Promise.resolve();
       }
       if (functionObject.package.individually || this.serverless.service.package.individually) {
         return this.packageFunction(functionName);
       }
       shouldPackageService = true;
-      return BbPromise.resolve();
-    });
-    const allLayers = this.serverless.service.getAllLayers();
-    packagePromises = packagePromises.concat(
-      allLayers.map(layerName => {
-        const layerObject = this.serverless.service.getLayer(layerName);
-        layerObject.package = layerObject.package || {};
-        if (layerObject.package.artifact) {
-          return BbPromise.resolve();
+      return Promise.resolve();
+    };
+
+    const packageLayer = layerName => {
+      const layerObject = this.serverless.service.getLayer(layerName);
+      layerObject.package = layerObject.package || {};
+      if (layerObject.package.artifact) {
+        return Promise.resolve();
+      }
+      return this.packageLayer(layerName);
+    };
+
+    return Promise.resolve()
+      .then(() => {
+        if (allFunctions.length > 0) {
+          this.serverless.cli.log(
+            `Packaging functions... (Up to ${MAX_CONCURRENT_PACKAGING} in parallel)`
+          );
+          return this._asyncPool(allFunctions, packageFunction, MAX_CONCURRENT_PACKAGING);
         }
-        return this.packageLayer(layerName);
+
+        return Promise.resolve();
       })
-    );
+      .then(() => {
+        if (allLayers.length > 0) {
+          this.serverless.cli.log(
+            `Packaging layers... (Up to ${MAX_CONCURRENT_PACKAGING} in parallel)`
+          );
+          return this._asyncPool(allLayers, packageLayer, MAX_CONCURRENT_PACKAGING);
+        }
 
-    return BbPromise.all(packagePromises).then(() => {
-      if (shouldPackageService && !this.serverless.service.package.artifact) {
-        return this.packageAll();
-      }
-      return BbPromise.resolve();
-    });
+        return Promise.resolve();
+      })
+      .then(() => {
```

Comment:
I know it's adding extra work to this PR, but I think in the first step we should convert these functions to use `async/await` and to make them not rely on Bluebird at all and then introduce this change.